### PR TITLE
[Part 1] Add support for cross-compiling s390x binaries

### PR DESCRIPTION
**Pull request by gajju26**
**What this PR does / why we need it**: The current release of Kubernetes support s390x architecture partly meaning we can build Kubernetes on s390x but while deploying Kubernetes it does not have required images pushed to gcr.io repository. This PR is to make Kubernetes docker images available for s390x architecture on gcr.io repository. This PR also enables s390x support to kube-dns , pause, addon-manager, etcd, hyperkube, kube-discovery etc.

**Which issue this PR fixes** #34328

**Special notes for your reviewer**: In existing file "build-tools/build-image/cross/Dockerfile" the repository mentioned for installing cross build tool chains for supporting architecture does not have a tool chain for s390x hence in my PR I am changing the repository so that it will be cross compiled for s390x.

**Release note**:

```release-note
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37092)

### CODE REVIEW

File: build-tools/build-image/cross/Dockerfile [hunk 1]
```diff
@@ -50,13 +52,11 @@ RUN mkdir -p /usr/local/src/protobuf \
...
-RUN echo "deb http://emdebian.org/tools/debian/ jessie main" > /etc/apt/sources.list.d/cgocrosscompiling.list \
-  && curl -s http://emdebian.org/tools/debian/emdebian-toolchain-archive.key | apt-key add - \
-  && for platform in ${KUBE_DYNAMIC_CROSSPLATFORMS}; do dpkg --add-architecture ${platform}; done \
+# To install crossbuild essential for other architectures add the following repository.
+RUN echo "deb http://archive.ubuntu.com/ubuntu xenial main universe" > /etc/apt/sources.list.d/cgocrosscompiling.list \
   && apt-get update \
-  && apt-get install -y build-essential \
-  && for platform in ${KUBE_DYNAMIC_CROSSPLATFORMS}; do apt-get install -y crossbuild-essential-${platform}; done \
+  && apt-get install -y --force-yes  build-essential \
+  && for platform in ${KUBE_DYNAMIC_CROSSPLATFORMS}; do apt-get install -y --force-yes crossbuild-essential-${platform}; done \
```
Comment (luxas) [inline build-tools/build-image/cross/Dockerfile]: why --force-yes?

File: build-tools/build-image/cross/Dockerfile [hunk 2]
```diff
@@ -50,13 +52,11 @@ RUN mkdir -p /usr/local/src/protobuf \
   && protoc --version
 
 # Use dynamic cgo linking for architectures other than amd64 for the server platforms
-# More info here: https://wiki.debian.org/CrossToolchains
-RUN echo "deb http://emdebian.org/tools/debian/ jessie main" > /etc/apt/sources.list.d/cgocrosscompiling.list \
-  && curl -s http://emdebian.org/tools/debian/emdebian-toolchain-archive.key | apt-key add - \
-  && for platform in ${KUBE_DYNAMIC_CROSSPLATFORMS}; do dpkg --add-architecture ${platform}; done \
+# To install crossbuild essential for other architectures add the following repository.
+RUN echo "deb http://archive.ubuntu.com/ubuntu xenial main universe" > /etc/apt/sources.list.d/cgocrosscompiling.list \
```
Comment (luxas) [inline build-tools/build-image/cross/Dockerfile]: Strange, http://archive.ubuntu.com/ubuntu/dists/xenial/ shows only i386 and amd64, although the `Release` and `InRelease` files show the other architectures.

**Review by luxas — COMMENTED** [review]
FYI @david-mcmahon @thockin @pwittrock @jessfraz re a new arch

@ixdy PTAL

**Comment by luxas** [comment]
@k8s-bot ok to test

**Comment by luxas** [comment]
@gajju26 Please sign the linux foundation CLA

File: build-tools/build-image/cross/Dockerfile (hunk 1 above)
Comment (gajju26) [inline build-tools/build-image/cross/Dockerfile]: Hi luxas,

I was getting "keyserver timed out GPG public key error" for the newly added repository which I could resolve using --force-yes option.

**Comment by gajju26** [comment]
Hi luxas,
I have signed "the linux foundation CLA" with same github user id i.e. gajju26.

File: build-tools/build-image/cross/Dockerfile (hunk 1 above)
Comment (ixdy) [inline build-tools/build-image/cross/Dockerfile]: it looks like if you run `apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 40976EAF437D05B` before `apt-get update` you don't need the `--force-yes`.

File: build-tools/build-image/cross/Dockerfile [hunk 3]
```diff
@@ -21,12 +21,14 @@ ENV GOARM 6
 ENV KUBE_DYNAMIC_CROSSPLATFORMS \
   armel \
   arm64 \
+  s390x \
```
Comment (ixdy) [inline build-tools/build-image/cross/Dockerfile]: I think you have trailing spaces after the `\`.

File: build-tools/build-image/cross/Dockerfile [hunk 4]
```diff
@@ -21,12 +21,14 @@ ENV GOARM 6
...
   armel \
   arm64 \
+  s390x \  
   ppc64el
 
 ENV KUBE_CROSSPLATFORMS \
   linux/386 \
   linux/arm linux/arm64 \
   linux/ppc64le \
+  linux/s390x \
```
Comment (ixdy) [inline build-tools/build-image/cross/Dockerfile]: trailing spaces here, too

File: build-tools/build-image/cross/Dockerfile (hunk 3 above)
Comment (gajju26) [inline build-tools/build-image/cross/Dockerfile]: Removed

File: build-tools/build-image/cross/Dockerfile (hunk 4 above)
Comment (gajju26) [inline build-tools/build-image/cross/Dockerfile]: Removed.

**Comment by gajju26** [comment]
@ixdy : my environment has some proxy issue due to which I was getting timeout error for " apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 40976EAF437D05B" but after resolving the proxy issue the command is working now and the file is also updated.

**Comment by ixdy** [comment]
Basically LGTM. @gajju26 can you bump up the version # in `build-tools/build-image/cross/VERSION`? I'll check out your PR, push a new version, and then we can merge.

**Comment by luxas** [comment]
> bump up the version # in build-tools/build-image/cross/VERSION

Yes, please do

Also, please add `linux/s390x` in the `hack/lib/golang.sh` file

**Comment by luxas** [comment]
@k8s-bot build this

**Comment by gajju26** [comment]
Hi @luxas @ixdy 
bump up the version # in build-tools/build-image/cross/VERSION -> Since golang 1.7.5 is not released hence bumped it to 1.7.4-1. Let me know if it is wrong.

please add  linux/s390x  in the  hack/lib/golang.sh  file ->  I have covered this change in part2 PR #36050, Do you want me to modify hack/lib/golang.sh in this PR?

**Comment by ixdy** [comment]
No need to update `hack/lib/golang.sh` in this PR, since that's being done in #36050.

@gajju26 can you rebase this PR? it looks like your PR is based off of the 1.7.3 crossbuild, so we can't merge it as-is.

**Comment by luxas** [comment]
Ah, ok, hadn't checked the other PR...

Cool, but needs a rebase to current master

**Comment by ixdy** [comment]
yeah, rebase and I'll push the cross-build image and we can merge and move forward!

**Comment by gajju26** [comment]
@luxas @ixdy 
I have rebased it to latest.

**Comment by ixdy** [comment]
`gcr.io/google_containers/kube-cross:v1.7.4-1` pushed.

@k8s-bot build this
@k8s-bot test this

**Comment by zhengxilong** [comment]
@gajju26 Any plan to have s390x binary download available from kubernetes release page?

**Comment by jessfraz** [comment]
we could cross compile it, i wrote the patch that cross compiles the go
binaries ;)

**Comment by luxas** [comment]
@zhengxilong s390x binaries will appear when v1.6.0-alpha.1 is released

@jessfraz Do you know when that's gonna happen?

Until that, it should be as easy as running something like 
```bash
build-tools/run.sh /bin/bash -c "KUBE_BUILD_PLATFORMS=linux/s390x make"
# or
make release
```

**Pull request merged**